File: fm-dddrive/src/main/java/io/dddrive/core/ddd/model/RepositoryDirectory.java
package io.dddrive.core.ddd.model;

import io.dddrive.core.ddd.model.impl.RepositoryDirectoryImpl;
import io.dddrive.core.enums.model.Enumerated;
import io.dddrive.core.enums.model.Enumeration;

public interface RepositoryDirectory {

	static RepositoryDirectory getInstance() {
		return RepositoryDirectoryImpl.getInstance();
	}

	<A extends Aggregate> AggregateRepository<A> getRepository(Class<A> intfClass);

	<A extends Aggregate, P extends Part<A>> PartRepository<A, P> getPartRepository(Class<P> intfClass);

	<A extends Aggregate> AggregatePersistenceProvider<A> getPersistenceProvider(Class<A> intfClass);

//	<A extends Aggregate, P extends Part<A>> PartPersistenceProvider<A, P> getPartPersistenceProvider(Class<P> intfClass);

	<E extends Enumerated> Enumeration<E> getEnumeration(Class<E> enumClass);

	Enumeration<? extends Enumerated> getEnumeration(String module, String enumName);

}

File: fm-dddrive/src/main/java/io/dddrive/core/ddd/model/RepositoryDirectory.kt
package io.dddrive.core.ddd.model

import io.dddrive.core.ddd.model.impl.RepositoryDirectoryImpl
import io.dddrive.core.enums.model.Enumerated
import io.dddrive.core.enums.model.Enumeration

interface RepositoryDirectory {

	fun <A : Aggregate> getRepository(intfClass: Class<A>): AggregateRepository<A>

	fun <A : Aggregate, P : Part<A>> getPartRepository(intfClass: Class<P>): PartRepository<A, P>

	fun <A : Aggregate> getPersistenceProvider(intfClass: Class<A>): AggregatePersistenceProvider<A>

	fun <P : Part<*>> getPartPersistenceProvider(intfClass: Class<P>): PartPersistenceProvider<P>

	fun <E : Enumerated> getEnumeration(enumClass: Class<E>): Enumeration<E>

	fun getEnumeration(
		module: String,
		enumName: String,
	): Enumeration<out Enumerated>

	companion object {

		@JvmStatic
		val instance: RepositoryDirectory = RepositoryDirectoryImpl()
	}

}
